### PR TITLE
Support touchscreen events from skiko on mobile`s browsers

### PR DESCRIPTION
**Pull request by jershell**
## Proposed Changes

Hello. The patch [part of touchscreen support.](https://github.com/JetBrains/skiko/pull/850) and should be accept after accepting the changes in skiko.
I tested it in Safari and Google Chrome on mobile devices. 

### JS Target
Сhrome desktop - ok
Safari desktop - ok
Firefox desktop - ok

Сhrome mobile - ok
Safari mobile - ok

### WASM Target
Сhrome desktop - ok
Сhrome mobile - ok

https://github.com/JetBrains/compose-multiplatform-core/assets/1238301/e15f2139-9664-48d5-871f-ef45fa4f7033

## Issues Fixed

Fixes: The bug on [Web: LazyColumn is not scrollable on mobile browser](https://github.com/JetBrains/compose-multiplatform/issues/3491) being fixed

## Google CLA
Done

### CODE REVIEW

**Comment by jershell** [comment]
Hello, skiko is ready.  Can we have release skiko and apply this patch?

**Comment by jershell** [comment]
@eymar. Hello. Can you help me with the PR?

**Comment by eymar** [comment]
@jershell Sure. It looks good to me. We'll update skiko next week, and I'll try to run your change.

File: compose/mpp/demo/src/wasmJsMain/resources/index.html [hunk 1]
```diff
@@ -35,6 +37,6 @@
 
 <body>
     <h1>compose wasm web demo</h1>
-    <canvas id="canvas1" width="800" height="600"></canvas>
+    <canvas id="canvas1"></canvas>
```
Comment (Schahen) [inline compose/mpp/demo/src/wasmJsMain/resources/index.html]: what's the purpose of this change? I mean, the PR is about touch events - how does the viewport area affects this issue?

Comment (eymar) [reply compose/mpp/demo/src/wasmJsMain/resources/index.html]: I guess the sizes here were redundant and irrelevant, since CanvasBasedWindow resizes the canvas according to the window size by default.

File: compose/ui/ui/src/jsNativeMain/kotlin/androidx/compose/ui/native/ComposeLayer.jsNative.kt [hunk 2]
```diff
@@ -63,10 +66,13 @@ internal class ComposeLayer(
         }
 
         override fun onPointerEvent(event: SkikoPointerEvent) {
-            if (supportsMultitouch) {
```
Comment (Schahen) [inline compose/ui/ui/src/jsNativeMain/kotlin/androidx/compose/ui/native/ComposeLayer.jsNative.kt:66]: it seems like if we are moving in this direction - I mean inspecting the nature of incoming event to resolve to multitouch this variable - supportsMultitouch is redundanant completely

File: compose/ui/ui/src/jsNativeMain/kotlin/androidx/compose/ui/native/ComposeLayer.jsNative.kt [hunk 3]
```diff
@@ -63,10 +66,13 @@ internal class ComposeLayer(
...
+            if (event.pointers.firstOrNull()?.device == SkikoPointerDevice.TOUCH) {
+                if (scene.platformContext.inputModeManager.inputMode != InputMode.Touch) {
```
Comment (Schahen) [inline compose/ui/ui/src/jsNativeMain/kotlin/androidx/compose/ui/native/ComposeLayer.jsNative.kt:70]: Can you please clarify what's the reasoning behind this check? When in theory we'll get a touch event but will be not in touch input mode? I might miss something since I haven't dove into touch events for quite a while, but from first glance this looks like  an attempt to initialise something that may me make sense to initialise/configure somewhere else, but not in the event handler itself. 

@eymar what do you think?

File: compose/mpp/demo/src/jsMain/resources/index.html [hunk 4]
```diff
@@ -4,7 +4,9 @@
     <meta charset="UTF-8">
     <title>compose multiplatform web demo</title>
     <script src="skiko.js"> </script>
-    <link type="text/css" rel="stylesheet" href="styles.css">
+    <meta name="viewport"
```
Comment (Schahen) [inline compose/mpp/demo/src/jsMain/resources/index.html]: The same comment as with viewport. if this has nothing to do with touch events - let's keep it out of scope of this particular commit - however small (and justified) the changes are.

**Review by Schahen — COMMENTED** [review]
Thank you for taking care of the multitouch. I've left some comments and will look for responses.

File: compose/mpp/demo/src/wasmJsMain/resources/index.html (hunk 1 above)
Comment (Schahen) [reply compose/mpp/demo/src/wasmJsMain/resources/index.html]: @jershell but this is irrelevant to the touch events as well - I suggest just push us to push both of the html/layout related changes right after we'll push this PR dedicated to changes in touch as a separate commit.

File: compose/ui/ui/src/jsNativeMain/kotlin/androidx/compose/ui/native/ComposeLayer.jsNative.kt (hunk 3 above)
Comment (jershell) [reply compose/ui/ui/src/jsNativeMain/kotlin/androidx/compose/ui/native/ComposeLayer.jsNative.kt:70]: We can connect keyboard/mouse to mobile device. And next event should change method from multi to single for hovering as example.

**Comment by eymar** [comment]
Apologies for a deleay with this PR. 
We decided to make a refactoring where skiko events handling is moved to Compose (so skiko's responsibility remains simply drawing on the canvas. Similar refactoring was done for iOS too and web should follow). Your skiko changes will be applied as well. And this PR will be applied too.

**Comment by jershell** [comment]
> Apologies for a deleay with this PR. We decided to make a refactoring where skiko events handling is moved to Compose (so skiko's responsibility remains simply drawing on the canvas. Similar refactoring was done for iOS too and web should follow). Your skiko changes will be applied as well. And this PR will be applied too.

I pulled latest changes from master. Touch events work. Its fine. But i have a couple of questions.
I removed from PR `<meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=0">` which broke the demo app(only mobile). I can to back this line here or in a separate PR. What shoud i do?
My next PR will be about ime\softwarekeyboard and depend on these changes. 
I think i could follow the plan

1. wait, the merger of this PR
2. make new PR for improve demo\js app
3. make new PR for IME Keyboard

alternative is connect second step to these or third changes

File: compose/ui/ui/src/jsNativeMain/kotlin/androidx/compose/ui/native/ComposeLayer.jsNative.kt (hunk 2 above)
Comment (jershell) [reply compose/ui/ui/src/jsNativeMain/kotlin/androidx/compose/ui/native/ComposeLayer.jsNative.kt:66]: I dont know. Yes, here it's look redundant, but i fear to remove `internal expect value ` from project could to bloat the PR. Can we make this moment as a topic on separate PR as nice refactoring of the future?

**Comment by eymar** [comment]
```<meta name="viewport" content="width=device-width, initial-scale=1.0">``` is necessary to support mobile browsers, but indeed a bit irrelevant to touch events. I think it was okay to keep it in this PR. Since you removed it already, I pushed this change to jb-main here - https://github.com/JetBrains/compose-multiplatform-core/pull/1061


I left only `width=device-width, initial-scale=1.0` because of these guidelines:

https://developer.mozilla.org/en-US/docs/Web/HTML/Viewport_meta_tag#maximum-scale
> Any value less than 3 fails accessibility

https://developer.mozilla.org/en-US/docs/Web/HTML/Viewport_meta_tag#user-scalable
> Setting the value to 0, which is the same as no, is against Web Content Accessibility Guidelines (WCAG).


Zooming doesn't work well in the demo (the app stops working). But it's a bug we'll need to solve and better to keep it on the surface.
___

The change looks good to me. It affects iOS too, so I'll check it works as before, and then we'll merge it!

**Review by eymar — APPROVED** [review]
iOS (tested on simulator) works too.

Thank you!
